restli-server/src/main/java/com/linkedin/restli/internal/server/response/BatchFinderResponseBuilder.java

```diff
@@ -195,26 +196,52 @@ private AnyRecord buildMetaData(CollectionResult<RecordTemplate, RecordTemplate>
     return null;
   }
 
-  private String getBatchParamterName(RoutingResult routingResult)
+  private String getBatchParameterName(RoutingResult routingResult)
   {
     int batchFinderCriteriaIndex = routingResult.getResourceMethod().getBatchFinderCriteriaParamIndex();
     return routingResult.getResourceMethod().getParameters().get(batchFinderCriteriaIndex).getName();
   }
 
   private DataList getCriteriaParameters(RoutingResult routingResult)
   {
-    String batchParamName = getBatchParamterName(routingResult);
-    return(DataList)routingResult.getContext().getStructuredParameter(batchParamName);
+    String batchParameterName = getBatchParameterName(routingResult);
+    return(DataList)routingResult.getContext().getStructuredParameter(batchParameterName);
   }
 
-  private URI buildCriteriaURI(ResourceContextImpl resourceContext, RecordTemplate criteria, String queryParam, URI uri)
+  private URI buildCriteriaURI(ResourceContextImpl resourceContext, RecordTemplate criteria, String batchParamName, URI uri)
   {
     DataList criteriaList = new DataList(1);
     criteriaList.add(criteria.data());
+    DataMap queryParams = extractQueryParamsFromResourceContext(resourceContext);
     return URIParamUtils.replaceQueryParam(uri,
-                                           queryParam,
+                                           batchParamName,
                                            criteriaList,
-                                           resourceContext.getParameters(),
+                                           queryParams,
                                            resourceContext.getRestliProtocolVersion());
   }
+
+  private DataMap extractQueryParamsFromResourceContext(ResourceContextImpl resourceContext)
+  {
+    try
+    {
+      DataMap queryParams = resourceContext.getParameters().copy();
```

Comment:
Deep copy might be expensive and for this usecase, a shallow copy is sufficient.